Feature/grove 424

maeisabelle opened the pull request:
This is now ready for code review and merging:

Changes include following the suggested approach in https://github.com/derms/ml-gradle/tree/dev/examples/disconnected-project-using-plugins-and-gradlew with minor changes. I also updated the versions of some gradle dependencies to latest.

I also tested changing the new gradle properties I added for offline use (for choosing the gradle home, offline directory, generated zip directory and filename, etc). I also tested trying to deploy the zip from a different directory and deploying offline from the unzipped package.

maeisabelle08 committed:
GROVE-424 - Pre-cook offline gradle use in grove-ml-gradle

maeisabelle08 committed:
GROVE-424 - Added note for changing projectGradleHome in gradle.properties

maeisabelle08 committed:
Updated commons-csv required by mlcp 9.0.9, updated the task makeOfflineZip to use the gradle properties for offline use and update README

grtjn commented:
Great stuff so far!

Couple of small details:

- I see you touched gradle-wrapper.. did you just pull in a newer version, or did you have to make changes to it yourself?
- I'd like to get rid of -Pdisconnected=true, and use the official gradle arg --offline instead. I think I used it in my attempts, which I shared in comments. If not, ping me offline..
- provide zip base name only, and append timestamp and/or version number (from package.json?), so you can use it to build a release package :)

maeisabelle08 committed:
*Replaced -Pdisconnected with --offline option for offline deployment
*Specified base zip file name instead of the zip filename with extension
*Appended timestamp (`yyyyMMddHHmmss`) offline zip to the project zip filename

maeisabelle commented:
For the first bullet, I followed https://github.com/derms/ml-gradle/blob/dev/examples/disconnected-project-using-plugins-and-gradlew/gradle/wrapper/gradle-wrapper.properties. If not, it will try to download the gradle library if the gradle.user.home is set to relative path. Otherwise, there will be no problem.

Working on the 2nd bullets and 3rd bullet. 

> Great stuff so far!
> 
> Couple of small details:
> 
> * I see you touched gradle-wrapper.. did you just pull in a newer version, or did you have to make changes to it yourself?
> * I'd like to get rid of -Pdisconnected=true, and use the official gradle arg --offline instead. I think I used it in my attempts, which I shared in comments. If not, ping me offline..
> * provide zip base name only, and append timestamp and/or version number (from package.json?), so you can use it to build a release package :)

grtjn commented:
Ah, good to know. Perhaps good to dig up some docs on the wrapper to make sure we follow official guidelines..

And no rush!

grtjn commented:
On second thought, can you support both --offline and -Pdisconnected=true. The benefit of the latter would be that you could use local props to pin down a certain copy to disconnected use permanently..

grtjn commented:
FYI: I have run tests with this on a real offline project, and seemed pretty much spot on. I need to compare notes to make sure though (might take some time)..

grtjn commented:
changed base to development..

grtjn commented:
hoping to find time to review this PR next week..

grtjn commented:
Gained a lot of experience from first hand with this in a project. I should really make time to take a close look at this, to polish a couple of things. Most just to make sure it works equally well for both online and offline cases.

grtjn commented on gradlew:
```diff
@@ -26,18 +26,21 @@ cd "$SAVED" >/dev/null
 APP_BASE_NAME=`basename "$0"`
+# Set the gradle home to be a local project directorys
+GRADLE_USER_HOME="$APP_HOME/build/gradle-home"
+echo "Setting the GRADLE_USER_HOME to $GRADLE_USER_HOME \n(needed for the creation of the offline repository)\n"
```
This doesn't seem to do the trick anymore.. I tried using -Dgradle.user.home near the end of the script instead

grtjn commented on gradlew:
```diff
@@ -155,14 +156,14 @@ if $cygwin ; then
 
 # Collect all arguments for the java command, following the shell quoting and substitution rules
-eval set -- $DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS "\"-Dorg.gradle.appname=$APP_BASE_NAME\"" -classpath "\"$CLASSPATH\"" org.gradle.wrapper.GradleWrapperMain "$APP_ARGS"
+eval set -- $DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS "\"-Dorg.gradle.appname=$APP_BASE_NAME\"" "\"-Dgradle.user.home=$GRADLE_USER_HOME\"" -classpath "\"$CLASSPATH\"" org.gradle.wrapper.GradleWrapperMain "$APP_ARGS"
```
ah. -D already there. Seems redundant to have it twice?

grtjn commented on gradlew.bat:
```diff
@@ -12,6 +12,8 @@ set DIRNAME=%~dp0
 set APP_BASE_NAME=%~n0
 set APP_HOME=%DIRNAME%
+set GRADLE_USER_HOME="%APP_HOME%\build\gradle-home"
+echo INFO: Set GRADLE_USER_HOME to %GRADLE_USER_HOME% (needed for the creation of the offline repository)
```
see comments gradlew

grtjn commented on build.gradle:
```diff
@@ -36,6 +47,51 @@ dependencies {
+  includeEmptyDirs false
+}
+
+def projectZipFilename = projectBaseZipFilename + '-' + new Date().format('yyyyMMddHHmmss') + '.zip'
```
I inserted an extra `project.name + '-' +`, as there might be multiple gradle projects you'd like/need to share..

grtjn commented on build.gradle:
```diff
@@ -36,6 +47,51 @@ dependencies {
 
+/**
+ * START: Disconnected gradle tasks
+ * Reference: https://github.com/derms/ml-gradle/blob/dev/examples/disconnected-project-using-plugins-and-gradlew/build.gradle
```
it was merged into ml-com: https://github.com/marklogic-community/ml-gradle/tree/master/examples/disconnected-project-using-plugins-and-gradlew

grtjn commented on build.gradle:
```diff
@@ -36,6 +47,51 @@ dependencies {
+ * START: Disconnected gradle tasks
+ * Reference: https://github.com/derms/ml-gradle/blob/dev/examples/disconnected-project-using-plugins-and-gradlew/build.gradle
+ */
+gradle.taskGraph.whenReady { graph ->
```
this, and following tasks only need to be available when 'online'..

grtjn commented on build.gradle:
```diff
@@ -7,13 +7,22 @@
+    jcenter()
+    mavenCentral()
+    maven { url "http://developer.marklogic.com/maven2/" }
+  }
```
this bit needs to be in buildscripts I think, and only there seems enough. Perhaps in case you need to compile something...

grtjn replied:
Nope, you need both places. Not entirely clear what serves what exactly. buildscripts seems to cater task deps, repositories and dependencies seems to cater compilation of code and such..

grtjn commented on settings.gradle:
```diff
@@ -0,0 +1,11 @@
+            gradlePluginPortal()
+        }
+    }
+}
```
crucial bit to make saliman plugin work in offline mode..

grtjn replied:
Also note it looks for startParameter.isOffline, not for disconnected property. Env properties not loaded yet at that stage, so you'd have to provide it from commandline (using -Pdisconnected=true), which kinda beats the purpose of using a property..

grtjn commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,5 @@
+distributionPath=build/offline/gradle/wrapper/dists
+zipStoreBase=PROJECT
+zipStorePath=build/offline/gradle/wrapper
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.8-bin.zip
```
gradlew kept insisting on downloading it from internet. seems like it looks in gradle-home for dists, despite the PROJECT setting.. I resorted to changing url to just the zip file name, and dropping a copy of the zip next to gradle/wrapper/gradle-wrapper.jar
